File: 2021/2021W29/PreppindataW29.py
import pandas as pd
import re

xlsx = pd.ExcelFile("F:/Data/Olympic Events.xlsx")
sport_pt = {'Artistic Gymnastics':'^Artistic Gymnastic.*', 'Baseball/Softball':'^(Baseball|Softball).*', 'Beach Volleyball':'^Beach Volley.*',
            'Boxing':'^Boxing.*', 'Rugby':'^Rugby.*', 'Skateboarding':'^Skateboarding.*', 'Wrestling':'^Wrestling.*'}
events = pd.read_excel(xlsx,'Olympics Events')
events['Date'] = events['Date'].apply(lambda x: re.sub('(\d+).+_(\w+)_(\d+)','\\1 \\2 \\3',x)) 
events['UK Date Time'] = events.apply(lambda x: x['Date']+' '+('0:00' if x['Time']=='xx' else x['Time']), axis=1)
events['Date'] = events['Date'].apply(lambda x: pd.to_datetime(x, format='%d %B %Y').date())
events['UK Date Time'] = events['UK Date Time'].apply(lambda x: pd.to_datetime(x, format='%d %B %Y %H:%M'))
for correct, pattern in sport_pt.items():
    events['Sport'] = events['Sport'].str.title().replace(to_replace = pattern, value = correct, regex = True)

events = pd.concat([events[['Date', 'UK Date Time', 'Sport', 'Venue']], 
                   pd.DataFrame([map(str.strip, x) \
                                 for x in events['Events'].str.split(',').values.tolist()])],
                  axis=1, sort=False)
events = events.melt(id_vars=[c for c in events.columns if not re.match('^\d', str(c))],
                          value_name='Event', var_name='ToDrop')\
                              .drop('ToDrop', axis=1).dropna(subset=['Event']).reset_index(drop=True)
events['Medal Ceremony?'] = events['Event'].apply(lambda x: bool(re.search('Victory Ceremony', x)) or bool(re.search('Gold Medal', x)))  
events['Venue'] =  events['Venue'].str.title()                            

venues = pd.read_excel(xlsx,'Venues')[['Venue', 'Location']].drop_duplicates()
temp = venues['Location'].str.split(',', n = 1, expand = True) 
venues['Latitude'] = temp[0].astype('float') 
venues['Longitude'] = temp[1].astype('float') 
venues = venues.drop('Location', axis=1)
venues['Venue'] =  venues['Venue'].str.title()   

final = pd.merge(events, venues, on='Venue')
final = final[['UK Date Time', 'Date', 'Sport', 'Event', 'Medal Ceremony?', 'Venue', 'Latitude', 'Longitude']]

File: 2021/2021W29/preppindataW29.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(stringi)
library(splitstackshape)

Sys.setlocale("LC_ALL","English")

xlsx <- "F:/Data/Olympic Events.xlsx"
final <- read_excel(path = xlsx, sheet = 'Olympics Events') %>%
  mutate('Date' = str_replace_all(`Date`, '(\\d+).+_(\\w+)_(\\d+)', '\\1 \\2 \\3'),
         'UK Date Time' =  as.POSIXct(paste(`Date`, if_else(`Time`=='xx','0:00',`Time`)), '%d %B %Y %H:%M', tz=Sys.timezone()),
         'Date' = as.Date(`Date`, '%d %B %Y'),
         'Venue' = str_to_title(`Venue`),
         'Sport' = stri_replace_all_regex(str_to_title(`Sport`), 
                                          pattern = c('^Artistic Gymnastic.*', '^(Baseball|Softball).*', '^Beach Volley.*',
                                                      '^Boxing.*', '^Rugby.*', '^Skateboarding.*', '^Wrestling.*'), 
                                          replacement = c('Artistic Gymnastic', 'Baseball/Softball', 'Beach Volleyball',
                                                          'Boxing', 'Rugby', 'Skateboarding', 'Wrestling'), 
                                          vectorize_all = FALSE)) %>%
  cSplit(., 'Events', ',') %>%
  pivot_longer(cols=starts_with('Events'), names_to='ToDrop', values_to='Event', values_drop_na = TRUE) %>%
  mutate('Medal Ceremony?' = (str_detect(`Event`, 'Victory Ceremony') | str_detect(`Event`, 'Gold Medal'))) %>%
  merge(., read_excel(path = xlsx, sheet = 'Venues') %>%
          select(c('Venue', 'Location')) %>%
          mutate('Venue' = str_to_title(`Venue`)) %>%
          distinct() %>%
          separate(`Location`, c('Latitude','Longitude'), sep=',', convert=TRUE), by='Venue') %>%
  select(c('UK Date Time', 'Date', 'Sport', 'Event', 'Medal Ceremony?', 'Venue', 'Latitude', 'Longitude'))

View(final)
